### PR TITLE
Issue 22: IPv6

### LINKED ISSUE #22: IPv6
- Base spec: https://tools.ietf.org/html/rfc8200
- Protocol parameters: https://www.iana.org/assignments/ipv6-parameters/ipv6-parameters.xhtml

### PR DESCRIPTION
**Pull request by senier**
# Description

Add IPv6 and ICMPv6 message specifications. Generate protocol numbers from IANA online database.

Closes #22

# Review Checklist

### Implementation

- [x] Does this code change accomplish what it is supposed to do?
- [x] Can this solution be simplified?
- [x] Does this change add unwanted compile-time or run-time dependencies?
- [x] Was a framework, API, library, service used that should not be used?
- [x] Was a framework, API, library, service not used that could improve the solution?
- [x] Is the code at the right abstraction level?
- [x] Is the code modular enough?
- [x] Would you have solved the problem in a different way that is substantially better in terms of the code’s maintainability, readability, performance, security?
- [x] Does similar functionality already exist in the codebase? If so, why isn’t this functionality reused?
- [x] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code?

### Logic Errors and Bugs

- [x] Can you think of any use case in which the code does not behave as intended?
- [x] Can you think of any inputs or external events that could break the code?

### Error Handling and Logging

- [x] Is error handling done the correct way?
- [x] Should any logging or debugging information be added or removed?
- [x] Are error messages user-friendly?
- [x] Are there enough log events and are they written in a way that allows for easy debugging?
- [x] Is the style of error / logging messages consistent?

### Dependencies

- [x] If this change requires updates outside of the code, like updating the documentation, configuration, readme files, was this done?
- [x] Might this change have any ramifications for other parts of the system, or backward compatibility?

### Security and Data Privacy

- [x] Does this code open the software up for security vulnerabilities?
- [x] Does this code change reveal some secret information like keys, passwords, or usernames (e.g., checked-in credentials)?

### Performance

- [x] Do you think this code change will impact system performance in a negative way?
- [x] Do you see any potential to improve the performance of the code?

### Usability

- [x] Is the proposed solution well designed from a usability perspective?
- [x] Is the API well documented?
- [x] Is the API/UI intuitive to use?

### Testing

- [x] Does it have enough automated tests (unit/integration/system tests)?
- [x] Are there some test cases, input or edge cases that should be tested in addition?

### Readability

- [x] Was the code easy to understand?
- [x] Can the readability of the code be improved by smaller methods?
- [x] Can the readability of the code be improved by different function/method or variable names?
- [x] Is the code located in the right file/folder/package?
- [x] Do you think certain methods should be restructured to have a more intuitive control flow?
- [x] Is the data flow understandable?
- [x] Are there redundant comments?
- [x] Could some comments convey the message better?
- [x] Would additional comments make the code more understandable?
- [x] Could some comments be removed by making the code itself more readable?
- [x] Is there any commented out code?

### Experts Opinion

- [x] Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be committed?
- [x] Will this code change impact different teams?
- [x] Should they have a say on the change as well?

### CODE REVIEW

File: icmpv6.rflx [hunk 1]
```diff
@@ -0,0 +1,105 @@
+package ICMPv6 is
+
+   type Reserved_8 is range 0 .. 0 with Size => 8;
```
Comment (treiher) [inline icmpv6.rflx:3]: What do you think of adding the name or a link to the RFC to each package?

**Review by treiher — CHANGES_REQUESTED** [review]

File: tools/generate_iana.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env -S python3 -O
+
+import re
```
[the remark this answers is not in the document]
Comment (senier) [reply tools/generate_iana.py]: Added in 8f647529.

File: iana.rflx [hunk 3]
```diff
@@ -0,0 +1,676 @@
...
+      --
+      -- Source:
+      --   [RXB5]
+      Prot_GMTP => 100,
+
+      -- Ipsilon Flow Management Protocol
+      --
+      -- Source:
+      --   November 1995, 1997.
+      Prot_IFMP => 101,
```
[the remark this answers is not in the document]
Comment (senier) [reply iana.rflx]: The IANA XML does some indirect mapping of person IDs to persons. I found this unnecessary, but for sake of completeness I added resolution of people in 8f647529.

File: iana.rflx [hunk 4]
```diff
@@ -0,0 +1,676 @@
...
+--  Updated:   2021-01-08
+package IANA is
+   -- Protocol Numbers
+   type Protocol_Number is (
+
+      -- IPv6 Hop-by-Hop Option
+      --
+      -- Source:
+      --   https://tools.ietf.org/html/rfc8200
+      Prot_HOPOPT => 0,
```
[the remark this answers is not in the document]
Comment (senier) [reply iana.rflx]: Changed in 8f64752 and c69b930.

File: icmpv6.rflx [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+   with Size => 8;
+
+   type MTU is range 1280 .. 2**32 - 1 with Size => 32;
+   type Pointer is range 0 .. 2**32 - 1 with Size => 32;
+   type Identifier is mod 2**16;
+   type Sequence_Number is mod 2**16;
+
+   type Message is
+      message
+         Kind : Kind
```
[the remark this answers is not in the document]
Comment (senier) [reply icmpv6.rflx]: Changed in 601015bb42d639b308b131d4d4b5235f6084e75b.

File: icmpv6.rflx [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            then Code_Time_Exceeded
+               if Kind = Time_Exceeded
+            then Code_Parameter_Problem
+               if Kind = Parameter_Problem
+            then Code_Unused
+               if Kind = Packet_Too_Big
+                  or Kind = Echo_Request
+                  or Kind = Echo_Reply
+            then Data
+               with Size => Message'Size - (Kind'Last - Kind'First + 1)
```
[the remark this answers is not in the document]
Comment (senier) [reply icmpv6.rflx]: Changed in 601015bb.

File: tests/data/ipv6/README.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
+Sample sources:
+
+# Wireshark Foundation sample captures
+
+## Example of IPv6 traffic using Teredo for encapsulation
```
[the remark this answers is not in the document]
Comment (senier) [reply tests/data/ipv6/README.md]: Changed in 601015bb.

File: ipv6_address.rflx [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+      message
+         null
+            then Data
+               with Size => 128;
+         Data : Opaque;
+      end message;
+
+   type Addresses is array of Address;
+
+end IPv6_Address;
```
[the remark this answers is not in the document]
Comment (senier) [reply ipv6_address.rflx]: Changed in cdbe928.

File: ipv6.rflx [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+   type Reserved_20 is range 0 .. 0 with Size => 20;
+   type Reserved_32 is range 0 .. 0 with Size => 32;
+
+   --  Options (RFC 8200, 4.2)
+
+   type Option_Action is
+      (Skip                                        => 0,
+       Discard                                     => 1,
+       Discard_With_Parameter_Problem              => 2,
+       Discard_With_Parameter_Problem_If_Multicast => 3) with Size => 2;
```
[the remark this answers is not in the document]
Comment (senier) [reply ipv6.rflx]: Fixed in 601015bb.

File: ipv6.rflx [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+         Traffic_Class : Traffic_Class;
+         Flow_Label : Flow_Label;
+         Payload_Length : Payload_Length;
+         Next_Header : IANA::Protocol_Number;
+         Hop_Limit : Hop_Limit;
+         Source : IPv6_Address::Address;
+         Destination : IPv6_Address::Address
+            then Data
+               with Size => 8 * Payload_Length;
+         Data : Opaque;
```
[the remark this answers is not in the document]
Comment (senier) [reply ipv6.rflx]: Changed in 601015bb.

File: ipv6.rflx [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+         Kind : Option_Kind
+            then null
+               if Kind_Action = Skip and Kind_May_Change = False and Kind_Value = Pad1
+            then Len
+               if Kind_Action /= Skip or Kind_May_Change = True or Kind_Value /= Pad1;
+         Len : Data_Len
+            then Data
+               with Size => 8 * Len;
+         Data : Opaque;
+      end message;
```
[the remark this answers is not in the document]
Comment (senier) [reply ipv6.rflx]: Changed in 601015bb.

File: ipv6.rflx [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+
+   type Options_Base_Header is
+      message
+         Next_Header : IANA::Protocol_Number;
+         Hdr_Ext_Len : Data_Len;
+         Options : Options
+            with Size => 48 + 64 * Hdr_Ext_Len;
+         Data : Opaque
+            with Size => Message'Last - Options'Last;
+      end message;
```
[the remark this answers is not in the document]
Comment (senier) [reply ipv6.rflx]: Documentation added in 601015bb.

File: ipv6.rflx [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+
+   type Segments_Count is range 0 .. 2**8 - 1 with Size => 8;
+
+   type Routing_Base_Header is
+      message
+         Hdr_Ext_Len : Data_Len;
+         Routing_Type : Routing_Type;
+         Data : Opaque
+            with Size => 40 + 64 * Hdr_Ext_Len;
+      end message;
```
[the remark this answers is not in the document]
Comment (senier) [reply ipv6.rflx]: Documentation added in 601015bb.

File: ipv6.rflx [hunk 14]
```diff
@@ -0,0 +1,368 @@
...
+   --  RPL Routing Header (RFC 6554, 3)
+
+   type Cmpr_Length is range 0 .. 15 with Size => 4;
+   type Pad_Length is range 0 .. 15 with Size => 4;
+
+   --  ISSUE: Componolit/systematization-binary-schemas#57
+   --  As per RFC 6554, section 3, the segments sent in the RPL routing header actually are a
+   --  vector of addresses. However, these are not fixed-size IPv6 addresses, but they are
+   --  truncated depending on the values of CmprI and CmprE. We cannot specify this as a normal
+   --  sequence because those require a static element type.
```
[the remark this answers is not in the document]
Comment (senier) [reply ipv6.rflx]: Yes, Componolit/RecordFlux#608.

File: ipv6.rflx [hunk 15]
```diff
@@ -0,0 +1,368 @@
...
+
+   type Segment_Routing_Header is
+      message
+         Segments_Left : Segments_Count;
+         Last_Entry : SRH_Last_Entry_Index;
+         Flags : Reserved_8;
+         Tag : SRH_Tag;
+         Segments : IPv6_Address::Addresses
+            with Size => 128 * (Last_Entry + 1);
+      end message;
```
[the remark this answers is not in the document]
Comment (senier) [reply ipv6.rflx]: Documented in 601015bb.

File: icmpv6.rflx (hunk 1 above)
Comment (senier) [reply icmpv6.rflx]: We should do that. Changed in 601015bb.

**Review by treiher — APPROVED** [review]
